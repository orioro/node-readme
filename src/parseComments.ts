import { Transform } from 'readable-stream'
import { parse } from 'comment-parser/lib'
import vinylFs from 'vinyl-fs'
import { groupBy } from 'lodash'

export const vinylCommentStream = () => {
  return new Transform({
    objectMode: true,
    transform(file, enc, cb) {
      if (file.isStream()) {
        this.emit('error', new Error('@orioro/readme: Streams files are not supported'))
        return cb()
      }

      if (file.isBuffer()) {
        file.comments = parse(file.contents.toString('utf8'), {
          spacing: 'preserve'
        })

        this.push(file)
      }

      cb()
    }
  })
}

const _parseComment = comment => {
  const { tags } = comment
  const nameTag = tags.find(tag => tag.tag === 'name') || null

  return {
    ...comment,
    name: nameTag ? nameTag.name : null,
    tags: groupBy(tags.filter(tag => tag.tag !== 'name'), 'tag'),
    description: typeof comment.description === 'string'
      ? comment.description.trim()
      : undefined
  }
}

/**
 * Parses comments loaded from files matching the given
 * glob patterns.
 * 
 * @name parseComments
 * @param {GlobPatterns[]} globs Array of glob patterns for files
 *                         from which comments should be loaded
 * @param {Object} options Options to be forwarded to `vinylFs.src`
 * @param {string} options.cwd
 * @return {Promise->Comment[]} List of comments
 */
export const parseComments = (
  globs,
  options?
) => (
  new Promise((resolve, reject) => {
    const comments = {}
    const commentsStream = vinylFs
      .src(globs, options)
      .pipe(vinylCommentStream())

    commentsStream.on('data', file => {
      comments[file.relative] = file.comments.reduce((acc, comment, index) => {
        comment = _parseComment(comment)

        const key = comment.name !== null ? comment.name : index + ''

        return {
          ...acc,
          [key]: comment
        }
      }, {})
    })

    commentsStream.on('end', () => resolve(comments))
    commentsStream.on('error', reject)
  })
)
