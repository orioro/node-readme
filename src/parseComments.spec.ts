import * as fs from 'fs'
import * as path from 'path'
import { parseComments } from './parseComments'

const FIXTURES_PATH = path.join(__dirname, '../test')
const TMP_PATH = path.join(__dirname, '../tmp')

describe('parseComments', () => {
  test('', () => {

    return parseComments('**/*.ts', {
      cwd: FIXTURES_PATH
    })
    .then(comments => {
      fs.writeFileSync(
        path.join(TMP_PATH, 'comments.json'),
        JSON.stringify(comments, null, '  '),
        'utf8'
      )
    })
  })
})
