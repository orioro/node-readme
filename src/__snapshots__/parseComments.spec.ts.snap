// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseComments  1`] = `
Object {
  "fixture-1/index.ts": Object {
    "func1": Object {
      "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit,
sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.

Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris
nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in
reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
pariatur.

\`\`\`js
const result = func1('some string', 10)
// false
\`\`\`

Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
      "name": "func1",
      "problems": Array [],
      "source": Array [
        Object {
          "number": 2,
          "source": "/**",
          "tokens": Object {
            "delimiter": "/**",
            "description": "",
            "end": "",
            "name": "",
            "postDelimiter": "",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": "",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 3,
          "source": " * Lorem ipsum dolor sit amet, consectetur adipiscing elit,",
          "tokens": Object {
            "delimiter": "*",
            "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit,",
            "end": "",
            "name": "",
            "postDelimiter": " ",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 4,
          "source": " * sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.",
          "tokens": Object {
            "delimiter": "*",
            "description": "sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.",
            "end": "",
            "name": "",
            "postDelimiter": " ",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 5,
          "source": " *",
          "tokens": Object {
            "delimiter": "*",
            "description": "",
            "end": "",
            "name": "",
            "postDelimiter": "",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 6,
          "source": " * Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris",
          "tokens": Object {
            "delimiter": "*",
            "description": "Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris",
            "end": "",
            "name": "",
            "postDelimiter": " ",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 7,
          "source": " * nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in",
          "tokens": Object {
            "delimiter": "*",
            "description": "nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in",
            "end": "",
            "name": "",
            "postDelimiter": " ",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 8,
          "source": " * reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla",
          "tokens": Object {
            "delimiter": "*",
            "description": "reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla",
            "end": "",
            "name": "",
            "postDelimiter": " ",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 9,
          "source": " * pariatur.",
          "tokens": Object {
            "delimiter": "*",
            "description": "pariatur.",
            "end": "",
            "name": "",
            "postDelimiter": " ",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 10,
          "source": " * ",
          "tokens": Object {
            "delimiter": "*",
            "description": "",
            "end": "",
            "name": "",
            "postDelimiter": " ",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 11,
          "source": " * \`\`\`js",
          "tokens": Object {
            "delimiter": "*",
            "description": "\`\`\`js",
            "end": "",
            "name": "",
            "postDelimiter": " ",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 12,
          "source": " * const result = func1('some string', 10)",
          "tokens": Object {
            "delimiter": "*",
            "description": "const result = func1('some string', 10)",
            "end": "",
            "name": "",
            "postDelimiter": " ",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 13,
          "source": " * // false",
          "tokens": Object {
            "delimiter": "*",
            "description": "// false",
            "end": "",
            "name": "",
            "postDelimiter": " ",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 14,
          "source": " * \`\`\`",
          "tokens": Object {
            "delimiter": "*",
            "description": "\`\`\`",
            "end": "",
            "name": "",
            "postDelimiter": " ",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 15,
          "source": " *",
          "tokens": Object {
            "delimiter": "*",
            "description": "",
            "end": "",
            "name": "",
            "postDelimiter": "",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 16,
          "source": " * Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
          "tokens": Object {
            "delimiter": "*",
            "description": "Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "end": "",
            "name": "",
            "postDelimiter": " ",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 17,
          "source": " *",
          "tokens": Object {
            "delimiter": "*",
            "description": "",
            "end": "",
            "name": "",
            "postDelimiter": "",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 18,
          "source": " * @name func1",
          "tokens": Object {
            "delimiter": "*",
            "description": "",
            "end": "",
            "name": "func1",
            "postDelimiter": " ",
            "postName": "",
            "postTag": " ",
            "postType": "",
            "start": " ",
            "tag": "@name",
            "type": "",
          },
        },
        Object {
          "number": 19,
          "source": " * @param {string} [strParam=''] Sed ut perspiciatis, unde omnis iste natus error",
          "tokens": Object {
            "delimiter": "*",
            "description": "Sed ut perspiciatis, unde omnis iste natus error",
            "end": "",
            "name": "[strParam='']",
            "postDelimiter": " ",
            "postName": " ",
            "postTag": " ",
            "postType": " ",
            "start": " ",
            "tag": "@param",
            "type": "{string}",
          },
        },
        Object {
          "number": 20,
          "source": " *                          sit voluptatem accusantium doloremque laudantium.",
          "tokens": Object {
            "delimiter": "*",
            "description": "sit voluptatem accusantium doloremque laudantium.",
            "end": "",
            "name": "",
            "postDelimiter": "                          ",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 21,
          "source": " * @param {number} [numParam=0] Nemo enim ipsam voluptatem, quia voluptas",
          "tokens": Object {
            "delimiter": "*",
            "description": "Nemo enim ipsam voluptatem, quia voluptas",
            "end": "",
            "name": "[numParam=0]",
            "postDelimiter": " ",
            "postName": " ",
            "postTag": " ",
            "postType": " ",
            "start": " ",
            "tag": "@param",
            "type": "{number}",
          },
        },
        Object {
          "number": 22,
          "source": " *                          sit, aspernatur aut odit aut fugit.",
          "tokens": Object {
            "delimiter": "*",
            "description": "sit, aspernatur aut odit aut fugit.",
            "end": "",
            "name": "",
            "postDelimiter": "                          ",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 23,
          "source": " * @return {boolean} result Nam libero tempore, cum soluta nobis est",
          "tokens": Object {
            "delimiter": "*",
            "description": "Nam libero tempore, cum soluta nobis est",
            "end": "",
            "name": "result",
            "postDelimiter": " ",
            "postName": " ",
            "postTag": " ",
            "postType": " ",
            "start": " ",
            "tag": "@return",
            "type": "{boolean}",
          },
        },
        Object {
          "number": 24,
          "source": " *                          eligendi optio.",
          "tokens": Object {
            "delimiter": "*",
            "description": "eligendi optio.",
            "end": "",
            "name": "",
            "postDelimiter": "                          ",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 25,
          "source": " */",
          "tokens": Object {
            "delimiter": "",
            "description": "",
            "end": "*/",
            "name": "",
            "postDelimiter": "",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
      ],
      "tags": Object {
        "param": Array [
          Object {
            "default": "''",
            "description": "Sed ut perspiciatis, unde omnis iste natus error
                         sit voluptatem accusantium doloremque laudantium.",
            "name": "strParam",
            "optional": true,
            "problems": Array [],
            "source": Array [
              Object {
                "number": 19,
                "source": " * @param {string} [strParam=''] Sed ut perspiciatis, unde omnis iste natus error",
                "tokens": Object {
                  "delimiter": "*",
                  "description": "Sed ut perspiciatis, unde omnis iste natus error",
                  "end": "",
                  "name": "[strParam='']",
                  "postDelimiter": " ",
                  "postName": " ",
                  "postTag": " ",
                  "postType": " ",
                  "start": " ",
                  "tag": "@param",
                  "type": "{string}",
                },
              },
              Object {
                "number": 20,
                "source": " *                          sit voluptatem accusantium doloremque laudantium.",
                "tokens": Object {
                  "delimiter": "*",
                  "description": "sit voluptatem accusantium doloremque laudantium.",
                  "end": "",
                  "name": "",
                  "postDelimiter": "                          ",
                  "postName": "",
                  "postTag": "",
                  "postType": "",
                  "start": " ",
                  "tag": "",
                  "type": "",
                },
              },
            ],
            "tag": "param",
            "type": "string",
          },
          Object {
            "default": "0",
            "description": "Nemo enim ipsam voluptatem, quia voluptas
                         sit, aspernatur aut odit aut fugit.",
            "name": "numParam",
            "optional": true,
            "problems": Array [],
            "source": Array [
              Object {
                "number": 21,
                "source": " * @param {number} [numParam=0] Nemo enim ipsam voluptatem, quia voluptas",
                "tokens": Object {
                  "delimiter": "*",
                  "description": "Nemo enim ipsam voluptatem, quia voluptas",
                  "end": "",
                  "name": "[numParam=0]",
                  "postDelimiter": " ",
                  "postName": " ",
                  "postTag": " ",
                  "postType": " ",
                  "start": " ",
                  "tag": "@param",
                  "type": "{number}",
                },
              },
              Object {
                "number": 22,
                "source": " *                          sit, aspernatur aut odit aut fugit.",
                "tokens": Object {
                  "delimiter": "*",
                  "description": "sit, aspernatur aut odit aut fugit.",
                  "end": "",
                  "name": "",
                  "postDelimiter": "                          ",
                  "postName": "",
                  "postTag": "",
                  "postType": "",
                  "start": " ",
                  "tag": "",
                  "type": "",
                },
              },
            ],
            "tag": "param",
            "type": "number",
          },
        ],
        "return": Array [
          Object {
            "description": "Nam libero tempore, cum soluta nobis est
                         eligendi optio.",
            "name": "result",
            "optional": false,
            "problems": Array [],
            "source": Array [
              Object {
                "number": 23,
                "source": " * @return {boolean} result Nam libero tempore, cum soluta nobis est",
                "tokens": Object {
                  "delimiter": "*",
                  "description": "Nam libero tempore, cum soluta nobis est",
                  "end": "",
                  "name": "result",
                  "postDelimiter": " ",
                  "postName": " ",
                  "postTag": " ",
                  "postType": " ",
                  "start": " ",
                  "tag": "@return",
                  "type": "{boolean}",
                },
              },
              Object {
                "number": 24,
                "source": " *                          eligendi optio.",
                "tokens": Object {
                  "delimiter": "*",
                  "description": "eligendi optio.",
                  "end": "",
                  "name": "",
                  "postDelimiter": "                          ",
                  "postName": "",
                  "postTag": "",
                  "postType": "",
                  "start": " ",
                  "tag": "",
                  "type": "",
                },
              },
              Object {
                "number": 25,
                "source": " */",
                "tokens": Object {
                  "delimiter": "",
                  "description": "",
                  "end": "*/",
                  "name": "",
                  "postDelimiter": "",
                  "postName": "",
                  "postTag": "",
                  "postType": "",
                  "start": " ",
                  "tag": "",
                  "type": "",
                },
              },
            ],
            "tag": "return",
            "type": "boolean",
          },
        ],
      },
    },
    "func2": Object {
      "description": "Aliquam at metus molestie, luctus arcu id, ornare ligula.",
      "name": "func2",
      "problems": Array [],
      "source": Array [
        Object {
          "number": 31,
          "source": "/**",
          "tokens": Object {
            "delimiter": "/**",
            "description": "",
            "end": "",
            "name": "",
            "postDelimiter": "",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": "",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 32,
          "source": " * Aliquam at metus molestie, luctus arcu id, ornare ligula.",
          "tokens": Object {
            "delimiter": "*",
            "description": "Aliquam at metus molestie, luctus arcu id, ornare ligula.",
            "end": "",
            "name": "",
            "postDelimiter": " ",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 33,
          "source": " *",
          "tokens": Object {
            "delimiter": "*",
            "description": "",
            "end": "",
            "name": "",
            "postDelimiter": "",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 34,
          "source": " * @name func2",
          "tokens": Object {
            "delimiter": "*",
            "description": "",
            "end": "",
            "name": "func2",
            "postDelimiter": " ",
            "postName": "",
            "postTag": " ",
            "postType": "",
            "start": " ",
            "tag": "@name",
            "type": "",
          },
        },
        Object {
          "number": 35,
          "source": " * @param {Object} param1",
          "tokens": Object {
            "delimiter": "*",
            "description": "",
            "end": "",
            "name": "param1",
            "postDelimiter": " ",
            "postName": "",
            "postTag": " ",
            "postType": " ",
            "start": " ",
            "tag": "@param",
            "type": "{Object}",
          },
        },
        Object {
          "number": 36,
          "source": " * @param {string} param1.opt1 Some option",
          "tokens": Object {
            "delimiter": "*",
            "description": "Some option",
            "end": "",
            "name": "param1.opt1",
            "postDelimiter": " ",
            "postName": " ",
            "postTag": " ",
            "postType": " ",
            "start": " ",
            "tag": "@param",
            "type": "{string}",
          },
        },
        Object {
          "number": 37,
          "source": " * @param {string} param1.opt2 Some other option",
          "tokens": Object {
            "delimiter": "*",
            "description": "Some other option",
            "end": "",
            "name": "param1.opt2",
            "postDelimiter": " ",
            "postName": " ",
            "postTag": " ",
            "postType": " ",
            "start": " ",
            "tag": "@param",
            "type": "{string}",
          },
        },
        Object {
          "number": 38,
          "source": " * @param {number} param2",
          "tokens": Object {
            "delimiter": "*",
            "description": "",
            "end": "",
            "name": "param2",
            "postDelimiter": " ",
            "postName": "",
            "postTag": " ",
            "postType": " ",
            "start": " ",
            "tag": "@param",
            "type": "{number}",
          },
        },
        Object {
          "number": 39,
          "source": " * @return {CustomType} result",
          "tokens": Object {
            "delimiter": "*",
            "description": "",
            "end": "",
            "name": "result",
            "postDelimiter": " ",
            "postName": "",
            "postTag": " ",
            "postType": " ",
            "start": " ",
            "tag": "@return",
            "type": "{CustomType}",
          },
        },
        Object {
          "number": 40,
          "source": " */",
          "tokens": Object {
            "delimiter": "",
            "description": "",
            "end": "*/",
            "name": "",
            "postDelimiter": "",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
      ],
      "tags": Object {
        "param": Array [
          Object {
            "description": "",
            "name": "param1",
            "optional": false,
            "problems": Array [],
            "source": Array [
              Object {
                "number": 35,
                "source": " * @param {Object} param1",
                "tokens": Object {
                  "delimiter": "*",
                  "description": "",
                  "end": "",
                  "name": "param1",
                  "postDelimiter": " ",
                  "postName": "",
                  "postTag": " ",
                  "postType": " ",
                  "start": " ",
                  "tag": "@param",
                  "type": "{Object}",
                },
              },
            ],
            "tag": "param",
            "type": "Object",
          },
          Object {
            "description": "Some option",
            "name": "param1.opt1",
            "optional": false,
            "problems": Array [],
            "source": Array [
              Object {
                "number": 36,
                "source": " * @param {string} param1.opt1 Some option",
                "tokens": Object {
                  "delimiter": "*",
                  "description": "Some option",
                  "end": "",
                  "name": "param1.opt1",
                  "postDelimiter": " ",
                  "postName": " ",
                  "postTag": " ",
                  "postType": " ",
                  "start": " ",
                  "tag": "@param",
                  "type": "{string}",
                },
              },
            ],
            "tag": "param",
            "type": "string",
          },
          Object {
            "description": "Some other option",
            "name": "param1.opt2",
            "optional": false,
            "problems": Array [],
            "source": Array [
              Object {
                "number": 37,
                "source": " * @param {string} param1.opt2 Some other option",
                "tokens": Object {
                  "delimiter": "*",
                  "description": "Some other option",
                  "end": "",
                  "name": "param1.opt2",
                  "postDelimiter": " ",
                  "postName": " ",
                  "postTag": " ",
                  "postType": " ",
                  "start": " ",
                  "tag": "@param",
                  "type": "{string}",
                },
              },
            ],
            "tag": "param",
            "type": "string",
          },
          Object {
            "description": "",
            "name": "param2",
            "optional": false,
            "problems": Array [],
            "source": Array [
              Object {
                "number": 38,
                "source": " * @param {number} param2",
                "tokens": Object {
                  "delimiter": "*",
                  "description": "",
                  "end": "",
                  "name": "param2",
                  "postDelimiter": " ",
                  "postName": "",
                  "postTag": " ",
                  "postType": " ",
                  "start": " ",
                  "tag": "@param",
                  "type": "{number}",
                },
              },
            ],
            "tag": "param",
            "type": "number",
          },
        ],
        "return": Array [
          Object {
            "description": "",
            "name": "result",
            "optional": false,
            "problems": Array [],
            "source": Array [
              Object {
                "number": 39,
                "source": " * @return {CustomType} result",
                "tokens": Object {
                  "delimiter": "*",
                  "description": "",
                  "end": "",
                  "name": "result",
                  "postDelimiter": " ",
                  "postName": "",
                  "postTag": " ",
                  "postType": " ",
                  "start": " ",
                  "tag": "@return",
                  "type": "{CustomType}",
                },
              },
              Object {
                "number": 40,
                "source": " */",
                "tokens": Object {
                  "delimiter": "",
                  "description": "",
                  "end": "*/",
                  "name": "",
                  "postDelimiter": "",
                  "postName": "",
                  "postTag": "",
                  "postType": "",
                  "start": " ",
                  "tag": "",
                  "type": "",
                },
              },
            ],
            "tag": "return",
            "type": "CustomType",
          },
        ],
      },
    },
  },
  "fixture-1/someMethod.ts": Object {
    "someMethod": Object {
      "description": "",
      "name": "someMethod",
      "problems": Array [],
      "source": Array [
        Object {
          "number": 2,
          "source": "/**",
          "tokens": Object {
            "delimiter": "/**",
            "description": "",
            "end": "",
            "name": "",
            "postDelimiter": "",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": "",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 3,
          "source": " * @name someMethod",
          "tokens": Object {
            "delimiter": "*",
            "description": "",
            "end": "",
            "name": "someMethod",
            "postDelimiter": " ",
            "postName": "",
            "postTag": " ",
            "postType": "",
            "start": " ",
            "tag": "@name",
            "type": "",
          },
        },
        Object {
          "number": 4,
          "source": " *",
          "tokens": Object {
            "delimiter": "*",
            "description": "",
            "end": "",
            "name": "",
            "postDelimiter": "",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
        Object {
          "number": 5,
          "source": " * @param {number} param Excepteur sint occaecat cupidatat non proident",
          "tokens": Object {
            "delimiter": "*",
            "description": "Excepteur sint occaecat cupidatat non proident",
            "end": "",
            "name": "param",
            "postDelimiter": " ",
            "postName": " ",
            "postTag": " ",
            "postType": " ",
            "start": " ",
            "tag": "@param",
            "type": "{number}",
          },
        },
        Object {
          "number": 6,
          "source": " * @return {number} result Lorem ipsum dolor sit amet, consectetur adipiscing",
          "tokens": Object {
            "delimiter": "*",
            "description": "Lorem ipsum dolor sit amet, consectetur adipiscing",
            "end": "",
            "name": "result",
            "postDelimiter": " ",
            "postName": " ",
            "postTag": " ",
            "postType": " ",
            "start": " ",
            "tag": "@return",
            "type": "{number}",
          },
        },
        Object {
          "number": 7,
          "source": " */",
          "tokens": Object {
            "delimiter": "",
            "description": "",
            "end": "*/",
            "name": "",
            "postDelimiter": "",
            "postName": "",
            "postTag": "",
            "postType": "",
            "start": " ",
            "tag": "",
            "type": "",
          },
        },
      ],
      "tags": Object {
        "param": Array [
          Object {
            "description": "Excepteur sint occaecat cupidatat non proident",
            "name": "param",
            "optional": false,
            "problems": Array [],
            "source": Array [
              Object {
                "number": 5,
                "source": " * @param {number} param Excepteur sint occaecat cupidatat non proident",
                "tokens": Object {
                  "delimiter": "*",
                  "description": "Excepteur sint occaecat cupidatat non proident",
                  "end": "",
                  "name": "param",
                  "postDelimiter": " ",
                  "postName": " ",
                  "postTag": " ",
                  "postType": " ",
                  "start": " ",
                  "tag": "@param",
                  "type": "{number}",
                },
              },
            ],
            "tag": "param",
            "type": "number",
          },
        ],
        "return": Array [
          Object {
            "description": "Lorem ipsum dolor sit amet, consectetur adipiscing",
            "name": "result",
            "optional": false,
            "problems": Array [],
            "source": Array [
              Object {
                "number": 6,
                "source": " * @return {number} result Lorem ipsum dolor sit amet, consectetur adipiscing",
                "tokens": Object {
                  "delimiter": "*",
                  "description": "Lorem ipsum dolor sit amet, consectetur adipiscing",
                  "end": "",
                  "name": "result",
                  "postDelimiter": " ",
                  "postName": " ",
                  "postTag": " ",
                  "postType": " ",
                  "start": " ",
                  "tag": "@return",
                  "type": "{number}",
                },
              },
              Object {
                "number": 7,
                "source": " */",
                "tokens": Object {
                  "delimiter": "",
                  "description": "",
                  "end": "*/",
                  "name": "",
                  "postDelimiter": "",
                  "postName": "",
                  "postTag": "",
                  "postType": "",
                  "start": " ",
                  "tag": "",
                  "type": "",
                },
              },
            ],
            "tag": "return",
            "type": "number",
          },
        ],
      },
    },
  },
}
`;
